File: ios/TruePadApp/TruePadAppUITests/FixedRecordSmokeTest.swift
import XCTest

/// THE FIXED-RECORD PROPERTY, OBSERVED ON THE HANDSET.
///
/// The engine's §16 behaviour is proven on the JVM and again in the Swift suite.
/// What is not proven there is that ticking a box in the real create flow produces
/// a pad whose messages all cost the same amount of pad — and that is the thing an
/// operator actually experiences.
///
/// WHY THE METERS ARE THE PROBE, not the transport string. The envelope on the
/// Send screen carries an explicit accessibility label — deliberately, so a screen
/// reader does not spell out several hundred characters of transport — which means
/// its text is not readable from the element tree. The pad's own "Pad material
/// left" row IS readable, and it is the better evidence anyway: it is the
/// operator-visible consequence of the property. Two plaintexts of very different
/// lengths must cost exactly the same number of bytes.
///
/// Every pad created here is DISPOSABLE and device-generated, so it reads NOT
/// ELIGIBLE and costs nothing. No operational pad is touched.
final class FixedRecordSmokeTest: XCTestCase {

    override func setUp() {
        super.setUp()
        continueAfterFailure = false
    }

    @discardableResult
    private func reveal(_ target: XCUIElement, in app: XCUIApplication, swipes: Int = 10) -> Bool {
        if target.waitForExistence(timeout: 3) { return true }
        for _ in 0..<swipes {
            app.swipeUp()
            if target.exists { return true }
        }
        for _ in 0..<(swipes * 2) {
            app.swipeDown()
            if target.exists { return true }
        }
        return target.exists
    }

    private func shoot(_ name: String) {
        let shot = XCTAttachment(screenshot: XCUIScreen.main.screenshot())
        shot.name = name
        shot.lifetime = .keepAlways
        add(shot)
    }

    /// "N bytes of pad material remain, out of M." -> N
    private func padMaterialRemaining(in app: XCUIApplication) -> Int? {
        let row = app.descendants(matching: .any)
            .matching(NSPredicate(format: "label CONTAINS %@", "bytes of pad material remain"))
            .firstMatch
        guard reveal(row, in: app) else { return nil }
        let label = row.label
        guard let r = label.range(of: "[0-9]+", options: .regularExpression) else { return nil }
        return Int(label[r])
    }

    /// Read the meter, opening "Security details" if it is not already showing.
    ///
    /// DELIBERATELY NOT DRIVEN BY THE DISCLOSURE'S REPORTED STATE. Keying on
    /// `value == "expanded"` made this depend on the very accessibility value the
    /// pass had just added, and on the disclosure being hittable at the moment it
    /// was asked — so a read could skip the tap and then find nothing. Looking for
    /// the meter first and only tapping when it is absent works whether the
    /// section is open, closed, or freshly rebuilt by a navigation.
    private func remainingBytes(in app: XCUIApplication) -> Int? {
        if let n = padMaterialRemaining(in: app) { return n }
        let details = app.buttons["Security details"]
        guard reveal(details, in: app) else { return nil }
        details.tap()
        return padMaterialRemaining(in: app)
    }

    private func send(_ message: String, in app: XCUIApplication) {
        let write = app.buttons["Write a message"]
        XCTAssertTrue(reveal(write, in: app), "the pad screen must offer Write a message")
        write.tap()

        let field = app.textViews["The message to send. It is encrypted with pad material "
                                  + "that is then destroyed."].firstMatch
        let target = field.exists ? field
            : app.textFields["The message to send. It is encrypted with pad material "
                             + "that is then destroyed."].firstMatch
        XCTAssertTrue(target.waitForExistence(timeout: 10), "the message field must render")
        target.tap()
        target.typeText(message)

        let encrypt = app.buttons["Encrypt and consume the pad"]
        XCTAssertTrue(reveal(encrypt, in: app), "the encrypt control must be reachable")
        encrypt.tap()

        // EXPLICIT CONTROLS FOR PUBLIC TRANSPORT MATERIAL.
        XCTAssertTrue(reveal(app.buttons["Copy"], in: app),
                      "the encrypted message has no explicit Copy control")
        XCTAssertTrue(app.buttons["Share"].exists,
                      "the encrypted message has no explicit Share control")
        XCTAssertTrue(reveal(app.buttons["Technical form"], in: app),
                      "the canonical JSON is no longer reachable")

        app.navigationBars.buttons.element(boundBy: 0).tap()
    }

    func testTwoMessagesOfDifferentLengthsCostTheSamePad() throws {
        let app = XCUIApplication()
        app.launch()
        XCTAssertTrue(app.wait(for: .runningForeground, timeout: 30))
        XCTAssertTrue(app.tabBars.buttons["Pads"].waitForExistence(timeout: 25))

        // ---- create a 256-byte fixed-record pad, through the real screen ----
        let create = app.buttons["Create a pad"]
        XCTAssertTrue(reveal(create, in: app))
        create.tap()

        let name = app.textFields["A name for this pad. It stays on this device."]
        XCTAssertTrue(name.waitForExistence(timeout: 10), "the create sheet must render")
        let label = "fixed-\(Int(Date().timeIntervalSince1970))"
        name.tap()
        name.typeText(label)

        // LENGTH PRIVACY LIVES UNDER "ADVANCED", collapsed by default.
        let advanced = app.buttons["Advanced"]
        XCTAssertTrue(reveal(advanced, in: app), "the Advanced disclosure must be reachable")
        advanced.tap()

        let hide = app.switches["Hide exact message lengths"]
        XCTAssertTrue(reveal(hide, in: app), "the fixed-length control is missing")
        // ASSERT THE TOGGLE ACTUALLY FLIPPED. A tap that lands on a SwiftUI
        // Toggle's label rather than its switch is silently a no-op, and the first
        // run of this test failed further down looking for a field that never
        // appeared — which reads as a missing control rather than an unflipped
        // switch. Checking the value turns that into a truthful failure.
        if hide.value as? String != "1" {
            hide.switches.firstMatch.exists ? hide.switches.firstMatch.tap() : hide.tap()
        }
        if hide.value as? String != "1" {
            hide.coordinate(withNormalizedOffset: CGVector(dx: 0.92, dy: 0.5)).tap()
        }
        XCTAssertEqual(hide.value as? String, "1",
                       "the fixed-length toggle did not turn on")
        shoot("01-create-fixed-length")

        // The field already offers 256, which is exactly what this pad wants.
        let size = app.textFields["Message size in bytes. Every message will use exactly "
                                  + "this much of the pad."]
        XCTAssertTrue(reveal(size, in: app), "the message size field did not appear")
        XCTAssertEqual(size.value as? String, "256",
                       "the offered fixed size is not the documented default")

        let submit = app.buttons["Create pad"]
        XCTAssertTrue(reveal(submit, in: app))
        XCTAssertTrue(submit.isEnabled, "Create is disabled with a valid 256-byte record size")
        submit.tap()

        // ---- open it ----
        let row = app.descendants(matching: .any)
            .matching(NSPredicate(format: "label CONTAINS %@", label)).firstMatch
        XCTAssertTrue(reveal(row, in: app, swipes: 12), "the new pad must appear in the list")
        row.tap()

        let readBefore = remainingBytes(in: app)
        if readBefore == nil { shoot("00-meter-unreadable") }
        let before = try XCTUnwrap(readBefore, "could not read the pad material meter")
        shoot("02-pad-detail-before")

        // ---- two plaintexts of clearly different lengths, both of which fit ----
        send("hi", in: app)
        let afterShort = try XCTUnwrap(remainingBytes(in: app))

        send("this one is a great deal longer than the other, on purpose", in: app)
        let afterLong = try XCTUnwrap(remainingBytes(in: app))
        shoot("03-pad-detail-after")

        let firstCost = before - afterShort
        let secondCost = afterShort - afterLong
        XCTAssertGreaterThan(firstCost, 0, "the first message consumed nothing")
        XCTAssertEqual(firstCost, secondCost,
                       "a two-character message and a fifty-eight-character message cost "
                       + "\(firstCost) and \(secondCost) bytes, so the fixed-record property does "
                       + "not hold on this device")
        XCTAssertEqual(firstCost, 256,
                       "each message cost \(firstCost) bytes, not the 256 that was asked for")
    }

    /// THE RECEIVE CODE HAS EXPLICIT CONTROLS TOO.
    ///
    /// Public transport material, like the encrypted message — and until this pass
    /// the iPhone offered neither, leaving the operator to select the text by hand.
    /// This publishes a request, checks the controls, and cancels it again; it does
    /// NOT perform a transfer and does not re-run the ceremony.
    func testTheReceiveCodeOffersCopyAndShare() throws {
        let app = XCUIApplication()
        app.launch()
        XCTAssertTrue(app.wait(for: .runningForeground, timeout: 30))
        app.tabBars.buttons["Inbox"].tap()

        // Start from a known state: drain anything a previous run published.
        var guardCount = 0
        while reveal(app.buttons["Cancel this code"], in: app, swipes: 4), guardCount < 6 {
            app.buttons["Cancel this code"].tap()
            guardCount += 1
        }

        let create = app.buttons["Create a receive code"]
        XCTAssertTrue(reveal(create, in: app), "the receive screen must offer a new code")
        create.tap()

        XCTAssertTrue(reveal(app.buttons["Copy code"], in: app),
                      "the receive code has no explicit Copy control")
        XCTAssertTrue(app.buttons["Share code"].exists,
                      "the receive code has no explicit Share control")
        shoot("04-receive-code-controls")

        // Leave the device as it was found.
        let cancel = app.buttons["Cancel this code"]
        if reveal(cancel, in: app) { cancel.tap() }
    }
}

File: android/app/src/androidTest/kotlin/dev/systemslibrarian/truepad/app/FixedRecordSmokeTest.kt
package dev.systemslibrarian.truepad.app

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.SystemClock
import androidx.compose.ui.test.assertIsDisplayed
import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onAllNodesWithTag
import androidx.compose.ui.test.onNodeWithTag
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import androidx.compose.ui.test.performScrollTo
import androidx.compose.ui.test.performTextClearance
import androidx.compose.ui.test.performTextInput
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import dev.systemslibrarian.truepad.core.COMPACT_PREFIX
import dev.systemslibrarian.truepad.core.EnvelopeDecode
import dev.systemslibrarian.truepad.core.decodeEnvelopeTransport2
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Rule
import org.junit.Test
import org.junit.rules.ExternalResource
import org.junit.rules.RuleChain
import org.junit.runner.RunWith
import java.io.File
import java.util.UUID

/**
 * THE FIXED-RECORD PROPERTY, OBSERVED ON A HANDSET.
 *
 * Everything about §16 records is already proven on the JVM. What is NOT proven
 * there is that the property survives the real screen: that ticking a checkbox in
 * the create flow actually produces a pad whose messages are all the same
 * ciphertext length, and that the string the operator is handed by the Copy button
 * is the compact form they were looking at.
 *
 * WHY THE CLIPBOARD IS THE PROBE. The envelope node carries
 * `clearAndSetSemantics`, so its text is deliberately invisible to the semantics
 * tree — a screen reader must not spell out 400 characters of transport. That
 * leaves the Copy button as the only way to observe what the screen is actually
 * offering, which is convenient: reading the clipboard tests the DISPLAYED value
 * and the COPY path at once. A test that read the envelope from the engine
 * instead would prove nothing about either.
 *
 * WHAT THIS DOES NOT COVER. It is one handset, so it says nothing about opening a
 * message written by the other edition. That needs the same pad on two devices.
 */
@RunWith(AndroidJUnit4::class)
class FixedRecordSmokeTest {

    private val context = InstrumentationRegistry.getInstrumentation().targetContext

    private val cleanStorage = object : ExternalResource() {
        override fun before() = wipe()
        override fun after() = wipe()
        private fun wipe() {
            AndroidStorage.storeRoot(context).deleteRecursively()
            AndroidStorage.witnessRoot(context).deleteRecursively()
            File(context.filesDir, "hidden-pads.txt").delete()
        }
    }

    private val compose = createAndroidComposeRule<MainActivity>()

    @get:Rule
    val chain: RuleChain = RuleChain.outerRule(cleanStorage).around(compose)

    private fun awaitTag(tag: String, timeoutMs: Long = 20_000) {
        compose.waitUntil(timeoutMs) { compose.onAllNodesWithTag(tag).fetchSemanticsNodes().isNotEmpty() }
    }

    /**
     * What is on the clipboard right now. `null` means the clipboard could not be
     * READ — no primary clip, or no item in it — which is a different thing from
     * an empty string, and on Android a very different thing: an app without
     * window focus is refused the read rather than given nothing. The timeout
     * message below says which of the two happened, because they have different
     * causes and only one of them is about TruePad.
     */
    private fun readClipboard(): String? {
        var text: String? = null
        InstrumentationRegistry.getInstrumentation().runOnMainSync {
            val cm = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            text = cm.primaryClip?.takeIf { it.itemCount > 0 }?.getItemAt(0)?.text?.toString()
        }
        return text
    }

    /**
     * Put a known value on the clipboard, so the wait below can tell "Copy
     * delivered this" apart from "this was already here".
     *
     * BEST EFFORT, DELIBERATELY. An instrumentation environment that refuses the
     * write is reported rather than fatal — the wait then falls back to its other
     * conditions, and the diagnostics say the seed did not take.
     */
    private fun seedClipboard(sentinel: String): Boolean =
        try {
            InstrumentationRegistry.getInstrumentation().runOnMainSync {
                val cm = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                cm.setPrimaryClip(ClipData.newPlainText("truepad-test-sentinel", sentinel))
            }
            readClipboard() == sentinel
        } catch (t: Throwable) {
            false
        }

    /**
     * Click the real Copy control and wait, WITH A DEADLINE, for the clipboard to
     * change into something that carries the compact transport.
     *
     * WHY THIS IS NOT ONE IMMEDIATE READ, which is what it was. `performClick`
     * returns when the click is DISPATCHED, not when the handler's clipboard write
     * has landed. On the CI emulator that read came back EMPTY, and the assertion
     * printed "Copy handed over something that is not the compact transport form:"
     * with nothing after the colon. An empty clipboard is what a read that arrives
     * before the write looks like; what the product actually put there in that run
     * was never observed, because the harness recorded nothing but the emptiness.
     * The same test passes on the physical handset and passed on this emulator
     * before and after, and the wait below now distinguishes the two cases rather
     * than leaving them to be inferred.
     *
     * AND WHY IT IS NOT AN UNBOUNDED WAIT, AND NOT "anything non-empty". A wait
     * with no deadline turns a genuinely broken Copy into a hung suite instead of
     * a red one. Accepting whatever happens to be on the clipboard would accept a
     * value a previous step left there — which is the very thing this test exists
     * to check. So the wait ends only on a value that is not the sentinel, is not
     * empty, and carries the compact prefix, and every assertion the test makes is
     * then made against THAT value.
     */
    private fun copyAndAwaitTransport(timeoutMs: Long = 15_000): String {
        val sentinel = "truepad-clipboard-sentinel-" + UUID.randomUUID()
        val seeded = seedClipboard(sentinel)

        // WHAT WAS THERE BEFORE THE CLICK — and the wait below requires the
        // clipboard to become something DIFFERENT from it.
        //
        // THIS, NOT THE SENTINEL, IS WHAT MAKES THE WAIT HONEST. This test calls
        // Copy TWICE, once per message. If the seed were the only guard and
        // seeding were refused by the platform, the second wait would find the
        // FIRST message still on the clipboard — non-empty, correctly prefixed,
        // and completely wrong — and return it as if Copy had delivered it. The
        // seed makes the "before" value predictable; comparing against the
        // "before" value is what makes a missing Copy detectable either way.
        val before = readClipboard()

        compose.onNodeWithTag("btn-copy-envelope").performClick()
        compose.waitForIdle()

        val deadline = SystemClock.uptimeMillis() + timeoutMs
        var last = before
        while (SystemClock.uptimeMillis() < deadline) {
            last = readClipboard()
            val value = last
            val changed = value != null && value.isNotEmpty() && value != before && value != sentinel
            if (changed && value.startsWith(COMPACT_PREFIX)) return value
            SystemClock.sleep(50)
        }

        // NEVER THE CONTENT. What sits on the clipboard here should be public
        // transport, but a harness that prints clipboards into a CI log is one
        // product defect away from printing a plaintext into a CI log. Report the
        // SHAPE of what was seen, which is what tells the failure modes apart.
        fail(
            "Copy did not deliver the compact transport within ${timeoutMs}ms. " +
                "sentinel seeded=$seeded, " +
                "clipboard unreadable=${last == null}, " +
                "clipboard unchanged since before the click=${last == before}, " +
                "clipboard still the sentinel=${last == sentinel}, " +
                "clipboard empty=${last?.isEmpty()}, " +
                "length=${last?.length}, " +
                "carries $COMPACT_PREFIX=${last?.startsWith(COMPACT_PREFIX)}",
        )
        error("unreachable")
    }

    private fun sendAndCopy(message: String): String {
        awaitTag("btn-send")
        compose.onNodeWithTag("btn-send").performClick()
        awaitTag("field-message")
        compose.onNodeWithTag("field-message").performTextInput(message)
        compose.onNodeWithTag("btn-encrypt").performScrollTo().performClick()
        awaitTag("envelope-output")

        // THE TRANSPORT FORM IS OFFERED UNDER A DISCLOSURE, and the compact form
        // is what is on screen. Both controls must be present before either is used.
        compose.onNodeWithTag("btn-copy-envelope").performScrollTo().assertIsDisplayed()
        compose.onNodeWithTag("btn-share-envelope").assertIsDisplayed()

        val copied = copyAndAwaitTransport()
        compose.onNodeWithTag("btn-back-to-pad").performScrollTo().performClick()
        return copied
    }

    @Test
    fun aFixedRecordPadGivesEveryMessageTheSameCiphertextLength() {
        // ---- create a 256-byte fixed-record pad, through the real screen ----
        awaitTag("btn-create-pad")
        compose.onNodeWithTag("btn-create-pad").performClick()
        awaitTag("field-pad-name")
        compose.onNodeWithTag("field-pad-name").performTextInput("fixed-smoke")
        compose.onNodeWithTag("size-Small").performScrollTo().performClick()

        // LENGTH PRIVACY IS UNDER "ADVANCED", COLLAPSED BY DEFAULT — which is the
        // point of putting it there, and is why the first run of this test could
        // not find the checkbox at all: a Details body is not merely off-screen
        // when closed, it is absent from the tree.
        compose.onNodeWithText("\u25B8  Advanced").performScrollTo().performClick()
        awaitTag("checkbox-fixed-length")
        compose.onNodeWithTag("checkbox-fixed-length").performScrollTo().performClick()
        awaitTag("field-fixed-size")
        // CLEARED FIRST. The field already offers 256, so typing into it produced
        // "256256" — which the screen then correctly refused as larger than a
        // Small pad's capacity, and the first run of this test timed out waiting
        // for a pad that the UI was right not to create.
        compose.onNodeWithTag("field-fixed-size").performTextClearance()
        compose.onNodeWithTag("field-fixed-size").performTextInput("256")
        compose.onNodeWithTag("btn-submit-create").performScrollTo().performClick()

        // CREATION OPENS THE PAD DIRECTLY. `createPadFromDevice` ends with
        // `openPad(...)`, so the home list is never shown and waiting for a
        // `pad-row` waits forever — the mistake the first runs of this test made.
        // Wait for the CONTROL, not a heading: the pad screen draws its title as
        // soon as a pad is selected, but its actions appear only once the engine
        // has returned a summary.
        awaitTag("btn-send", 60_000)

        // ---- two plaintexts of clearly different lengths, both of which fit ----
        val short = "hi"
        val long = "this one is a great deal longer than the other, on purpose"
        assertNotEquals("the fixtures must differ in length", short.length, long.length)

        val firstCopied = sendAndCopy(short)
        val secondCopied = sendAndCopy(long)

        // ---- what the operator was handed is the compact form ----
        //
        // REDUNDANT BY CONSTRUCTION, AND KEPT ANYWAY. `copyAndAwaitTransport` only
        // returns on a value carrying this prefix, so today this cannot fail. It
        // is the assertion that starts biting again the moment somebody loosens
        // the wait's exit condition — which is exactly the edit that would make
        // this test accept the wrong string.
        for (copied in listOf(firstCopied, secondCopied)) {
            assertTrue(
                "Copy handed over something that is not the compact transport form: " +
                    copied.take(40),
                copied.startsWith(COMPACT_PREFIX),
            )
        }
        assertNotEquals("the two messages produced identical transport", firstCopied, secondCopied)

        // ---- and the fixed-record property holds on the handset ----
        val a = decodeEnvelopeTransport2(firstCopied)
        val b = decodeEnvelopeTransport2(secondCopied)
        assertTrue("the first copied message does not decode", a is EnvelopeDecode.Ok)
        assertTrue("the second copied message does not decode", b is EnvelopeDecode.Ok)
        val lenA = (a as EnvelopeDecode.Ok).envelope.ciphertextLength
        val lenB = (b as EnvelopeDecode.Ok).envelope.ciphertextLength
        assertEquals(
            "two messages of different lengths produced different ciphertext lengths, so the " +
                "fixed-record property does not hold on this device",
            lenA,
            lenB,
        )
        assertEquals("the record is not the 256 bytes that was asked for", 256, lenA)
    }
}
